plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -1,221 +1,461 @@
+                pSampleBuffer += numberOfSamplesRead;
+                m_sampleBufferReadOffset += numberOfSamplesRead;
+            }
+            m_curFrameIndex += numberOfFramesRead;
+            DEBUG_ASSERT(isValidFrameIndex(m_curFrameIndex));
+            numberOfFramesRemaining -= numberOfFramesRead;
+            if (0 == numberOfFramesRemaining) {
+                break; // exit loop
+            }
+            // All previously decoded sample data has been consumed
+            DEBUG_ASSERT(m_sampleBufferReadOffset == m_sampleBufferWriteOffset);
+        }
+        m_sampleBufferReadOffset = 0;
+        m_sampleBufferWriteOffset = 0;
+
+        DEBUG_ASSERT(m_inputBufferOffset <= m_inputBufferLength);
+        if (m_inputBufferOffset >= m_inputBufferLength) {
+            // reset input buffer
+            m_inputBufferOffset = 0;
+            m_inputBufferLength = 0;
+            if (isValidSampleBlockId(m_curSampleBlockId)) {
+                // fill input buffer with next block of samples
+                u_int8_t* pInputBuffer = &m_inputBuffer[0];
+                u_int32_t inputBufferLength = m_inputBuffer.size(); // in/out parameter
+                if (!MP4ReadSample(m_hFile, m_trackId, m_curSampleBlockId,
+                        &pInputBuffer, &inputBufferLength,
+                        NULL, NULL, NULL, NULL)) {
+                    qWarning()
+                            << "Failed to read MP4 input data for sample block"
+                            << m_curSampleBlockId << "(" << "min ="
+                            << kSampleBlockIdMin << "," << "max ="
+                            << m_maxSampleBlockId << ")";
+                    break; // abort
+                }
+                ++m_curSampleBlockId;
+                m_inputBufferLength = inputBufferLength;
+            }
+        }
+        DEBUG_ASSERT(m_inputBufferOffset <= m_inputBufferLength);
+
+        // NOTE(uklotzde): The sample buffer for NeAACDecDecode2 has to
+        // be big enough for a whole block of decoded samples, which
+        // contains up to kFramesPerSampleBlock frames. Otherwise
+        // we need to use a temporary buffer.
+        CSAMPLE* pDecodeBuffer; // in/out parameter
+        SINT decodeBufferCapacityInBytes;
+        if (pSampleBuffer && (numberOfFramesRemaining >= kFramesPerSampleBlock)) {
+            // decode samples directly into sampleBuffer
+            pDecodeBuffer = pSampleBuffer;
+            decodeBufferCapacityInBytes = frames2samples(
+                    numberOfFramesRemaining) * sizeof(*pSampleBuffer);
+        } else {
+            // decode next block of samples into temporary buffer
+            pDecodeBuffer = &m_sampleBuffer[m_sampleBufferWriteOffset];
+            const int sampleBufferCapacity =
+                    m_sampleBuffer.size() -
+                    m_sampleBufferWriteOffset;
+            DEBUG_ASSERT(sampleBufferCapacity >=
```

Comment:
Why is this assertion true? 
.. because m_sampleBufferWriteOffset is allways 0 here!

I think we should not use m_sampleBufferWriteOffset here because it implicates a ring buffer and it is not used for writing to an offset.  
We should replace it with m_sampleBufferSampleCount or something, that just indicates the number of valid samples in the  buffer.